Basic configuration tests

Ryan089 opened the pull request:
## Summary
Creates a basic test script to confirm basic configuration of objects:
- Serialized fields have been set. The purpose is to prevent NullReferenceExceptions caused by failing to initialize fields.
- GameObjects in a particular scene are on the correct layer. The purpose is to ensure layer-based functionality (Physics, Raycasts, Rendering etc) works as expected.
- Prefabs are on the correct layer. The purpose is to ensure that layer-based functionality works correctly when new objects are added to scenes.

## Changes to Files
No changes to existing files. All new files are contained within Assets/EditorTests.

## Technical Notes
- The fields tested are configured within Assets/EditorTests/Test Data/DataSource.cs, and accessed using reflection. Separate test cases are run for each class tested.

- Changes to tested variable names will cause the test to fail. However, this 'false positive' will be caught when running the test manually or as part of the CI pipeline, and is easily remediated by amending DataSource.cs.

- Additional test cases are trivial to add, through modifying DataSource.cs.
 
## Suggested method for review

- [ ] Open the Unity Test Runner window (Window > General > Test Runner). Run the "SerializedFieldsAreNotNull" test in Unity Test Runner. All six test cases should pass.
![image](https://user-images.githubusercontent.com/64360820/181118291-ca31d028-712f-4ee3-985a-9233257754ee.png)

- [ ] Open the Lobby Scene, find GameObject "Chat". Delete the reference stored in the "Tab Button" field of the GenericTabView component. Re-run the test. The corresponding test case should fail, informing you of the null variable. The five other test cases should pass. Before proceeding, restore "Chat" to original condition and ensure test passes.
![image](https://user-images.githubusercontent.com/64360820/181116486-9a9d0c67-00a5-4a2a-8ff6-9d9781a7e2e4.png)

- [ ] Within DataSource.cs, add a typographical error into any of the variable names (such as "_roundCountdownText") in the NullCheckData() method. Re-run the test. Test should fail, informing you that the specified class does not have a variable of that name. Before proceeding, restore script to original condition and ensure test passes.
![image](https://user-images.githubusercontent.com/64360820/181117366-aef3daf9-a4e5-48c6-bdd9-88f580ef1f3c.png)

- [ ] Run the "GameObjectsAreOnTheirMandatedLayers" test in Unity Test Runner. Both test cases should pass.
![image](https://user-images.githubusercontent.com/64360820/181118347-e8f2f030-56cf-4c44-8b41-e8d66b1f6d8c.png)

- [ ] Open the Lobby Scene, find GameObject "Chat" (specifically, the one with the GenericTabView script). Change its layer in the inspector to anything other than UI. Re-run the test. Test should fail, informing you that the object is on the incorrect layer. Before proceeding, restore object to original condition and ensure test passes.
![image](https://user-images.githubusercontent.com/64360820/181117833-107821d6-0351-4c0a-8b5b-732360ecae91.png)

- [ ] Run the "PrefabsAreOnTheirMandatedLayers" test in Unity Test Runner. Both test cases should pass.
![image](https://user-images.githubusercontent.com/64360820/181118420-290e292a-3d68-4e10-b2f0-9394b52257fd.png)

- [ ] In the project browser, find prefab "PlayerNameUI". Change its layer in the inspector to anything other than UI. Re-run the test. Test should fail, informing you that the prefab is on the incorrect layer. Before proceeding, restore prefab to original condition and ensure test passes.
![image](https://user-images.githubusercontent.com/64360820/181118152-0c436968-f67f-4715-b12f-b9503fa86f91.png)

- [ ] Conduct static review of code to confirm adherence to programming guidelines, alignment to project test strategy etc.

Ryan089 committed:
Basic configuration tests to:
- Confirm required serialized variables have been initialized
- Confirm game objects in a particular scene are on the correct layer
- Confirm prefabs are on the correct layer

Ryan089 commented:
Closed as it isn't integrated into CI pipeline and is thus pointless.

Alainx277 reviewed (approved):
Didn't test it in Unity, idea and code are 👍 .

Ryan089 committed:
Substantially refactored configuration testing code.

Specifying not-null fields and mandatory layers no longer occurs within DataSource.cs. Rather, attributes have been created to decorate those fields / classes directly, and these are accessed in the test through reflection.

Ryan089 commented:
I have updated the structure of the code, removing the requirement for DataSource.cs and instead relying on custom attributes to identify items to test. Tests will delay failure until all checks are complete, so that the user has immediate access to a comprehensive list of breaches (rather than requiring multiple runs of the test). All three tests currently pass.
![image](https://user-images.githubusercontent.com/64360820/183660555-c99195dc-f4f9-469a-9ef8-075611b2c1ef.png)

Example of multiple breaches reported for a single test:
![image](https://user-images.githubusercontent.com/64360820/183660117-daef7922-bade-4e69-bf2f-28236d1add19.png)


Example syntax to mandate a specified layer for a class:
```cs
[RequiredLayer("UI")]
public sealed class GenericTabView : MonoBehaviour
```

Example syntax to mandate that a field must not be null:
```cs
[SerializeField] [NotNull] private Transform _panelUI;
```

Alainx277 commented on Assets/EditorTests/LobbyTests.cs:
```diff
@@ -0,0 +1,199 @@
+        {
+            prefabMonoBehaviours = prefab.GetComponentsInChildren<MonoBehaviour>();
+
+            foreach (MonoBehaviour mono in prefabMonoBehaviours)
+            {
+                Type monoType = mono.GetType();
+                RequiredLayerAttribute attribute = Attribute.GetCustomAttribute(monoType, typeof(RequiredLayerAttribute)) as RequiredLayerAttribute;
+                if (attribute != null)
+                {
+                    // Once we are here, we have found a MonoBehaviour with a RequiredLayerAttribute.
+                    // We now need to test the GameObject to see if it is on the layer that is mandated.
+
+                    if (mono.gameObject.layer != LayerMask.NameToLayer(attribute.layer))
+                    {
+                        // The test will fail, as the GameObject SHOULD have had been on a specific layer, but WAS NOT.
+                        // We are delaying the assertion so that all errors are identified in the console, rather than requiring the
+                        // test to be run multiple times (and only identifying a single breach each time).
+                        allRelevantMonoBehavioursAreOnTheRightLayer = false;
+                        sb.Append($"-> {monoType.Name} script requires object '{mono.gameObject.name}' (in prefab {prefab.name}) to be on {attribute.layer} layer, but it was on {LayerMask.LayerToName(mono.gameObject.layer)} layer.\n");
+                    }
+                }
+            }
```
This code is duplicated from the test above. Can you extract it?

Ryan089 replied:
Haha you caught me! I had the same thought when I wrote it, but didn't think anyone would notice / care. I've added a commit to tidy it up.

Ryan089 committed:
Extracted duplicate code from mandated layer checks.

Ryan089 committed:
Editor Tests moved into Scripts folder.
